src/parser/mod.rs

```diff
@@ -580,7 +583,7 @@ fn member_access<'a, 'b>(input: Input<'a, 'b>) -> AstResult<'a, 'b> {
 fn argument<'a, 'b>(input: Input<'a, 'b>) -> AstResult<'a, 'b> {
     match input[0].0 {
         Token::StringType => variable(input),
-        Token::Identifier(_) => variable(input),
+        Token::TypeName(_) | Token::Identifier(_) => or(&[variable, variant], "argument")(input),
```

Comment:
What is the reasoning for combining the TypeName/variant and Identifier/variable rules into one case?

Edit: Ah because of the shared path I see. Since paths can only consist of TypeNames I think it may be worth keeping the fast case for single variables `Identifier => variable`, and only combining them in the Typename case: `Typename => or([variable, variant])`. Also may be good to make module_path its own parser to avoid redoing work. Then this case can be some equivalent of `module_path then or(variant, variable)` (pseudocode) where neither of variant or variable internally use module_path anymore.